fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanHashable.java

```diff
@@ -31,31 +31,57 @@
 
 /**
  * A more stable version of {@link Object#hashCode}.
+ * The planHash semantics are different than {@link Object#hashCode} in a few ways:
+ * <UL>
+ *     <LI>{@link #planHash()} values should be stable across runtime instance changes. The reason is that these values can be used to validate
+ *     outstanding continuations, and a change in hash value caused by an application restart or refactoring will invalidate all those
+ *     outstanding continuations</LI>
+ *     <LI>{@link #planHash()} supports multiple flavors of hash calculations (See {@link PlanHashKind}). The various kinds of plan hash algorithms
+ *     are used for different purposes and include/exclude different parts of the target query plan</LI>
+ *     <LI>{@link #planHash()} is meant to imply a certain identity of a plan, and reflects on the entire structure of the plan.
+ *     The intent is to be able to correlate various plans for "identity" (using different definitions for this identity as
+ *     specified by {@link PlanHashKind}). This requirement drives a desire to reduce collisions as much as possible since
+ *     not in all cases can we actually use "equals" to verify identity (e.g. log messages)</LI>
+ * </UL>
  */
 @API(API.Status.UNSTABLE)
 public interface PlanHashable {
+    /**
+     * The "kinds" of planHash calculations.
+     */
+    enum PlanHashKind {
+        LEGACY,                       // The original plan hash kind. Here for backwards compatibility, will be removed in the future
+        FOR_CONTINUATION,             // Continuation validation plan hash kind: include children, literals and markers. Used for continuation validation
+        STRUCTURAL_WITHOUT_LITERALS   // The hash used for query matching: skip all literals and markers
+    }
+
     /**
      * Return a hash similar to <code>hashCode</code>, but with the additional guarantee that is is stable across JVMs.
+     * @param hashKind the "kind" of hash to calculate. Each kind of hash has a particular logic with regards to included and excluded items.
      * @return a stable hash code
      */
-    int planHash();
+    int planHash(@Nonnull final PlanHashKind hashKind);
+
+    default int planHash() {
```

Comment:
This is more of a worry than an actual concern. I would love to make this method `final` but Java does not support that (yet). Can you hunt down all overrides (I think you have already done that) and make sure they actually get changed into the `planHash(PlanHashKind)`. This method should not be overridden.

Reply:
Done